File: frontend/src/api/events.ts
import { api } from './client'
import type { Event, EventListResponse } from '../types'

export const eventsApi = {
  list: (slug: string, params?: { event_type_id?: string; search?: string; offset?: number; limit?: number }) => {
    const sp = new URLSearchParams()
    if (params?.event_type_id) sp.set('event_type_id', params.event_type_id)
    if (params?.search) sp.set('search', params.search)
    if (params?.offset !== undefined) sp.set('offset', String(params.offset))
    if (params?.limit !== undefined) sp.set('limit', String(params.limit))
    const qs = sp.toString()
    return api.get<EventListResponse>(`/projects/${slug}/events${qs ? `?${qs}` : ''}`)
  },
  get: (slug: string, id: string) => api.get<Event>(`/projects/${slug}/events/${id}`),
  create: (slug: string, data: {
    event_type_id: string; name: string; description?: string;
    field_values?: { field_definition_id: string; value: string }[];
    meta_values?: { meta_field_definition_id: string; value: string }[];
  }) => api.post<Event>(`/projects/${slug}/events`, data),
  update: (slug: string, id: string, data: {
    name?: string; description?: string;
    field_values?: { field_definition_id: string; value: string }[];
    meta_values?: { meta_field_definition_id: string; value: string }[];
  }) => api.patch<Event>(`/projects/${slug}/events/${id}`, data),
  del: (slug: string, id: string) => api.del(`/projects/${slug}/events/${id}`),
  bulkCreate: (slug: string, data: {
    event_type_id: string; name: string;
    field_values?: { field_definition_id: string; value: string }[];
    meta_values?: { meta_field_definition_id: string; value: string }[];
  }[]) => api.post<Event[]>(`/projects/${slug}/events/bulk`, data),
}

File: backend/src/tripl/api/v1/events.py
import uuid

from fastapi import APIRouter, Query

from tripl.api.deps import SessionDep
from tripl.schemas.event import (
    EventBulkDelete,
    EventCreate,
    EventListResponse,
    EventMove,
    EventResponse,
    EventUpdate,
)
from tripl.services import event_service

router = APIRouter(prefix="/projects/{slug}/events", tags=["events"])


@router.get("", response_model=EventListResponse)
async def list_events(
    session: SessionDep,
    slug: str,
    event_type_id: uuid.UUID | None = None,
    search: str | None = None,
    implemented: bool | None = None,
    tag: str | None = None,
    reviewed: bool | None = None,
    archived: bool | None = None,
    offset: int = Query(0, ge=0),
    limit: int = Query(200, ge=1, le=10000),
):
    items, total = await event_service.list_events(
        session, slug, event_type_id, search, implemented, tag, reviewed, archived, offset, limit
    )
    return EventListResponse(items=items, total=total)


@router.get("/tags", response_model=list[str])
async def list_tags(session: SessionDep, slug: str):
    return await event_service.list_tags(session, slug)


@router.post("", response_model=EventResponse, status_code=201)
async def create_event(session: SessionDep, slug: str, data: EventCreate):
    return await event_service.create_event(session, slug, data)


@router.post("/bulk", response_model=list[EventResponse], status_code=201)
async def bulk_create_events(session: SessionDep, slug: str, data: list[EventCreate]):
    return await event_service.bulk_create_events(session, slug, data)


@router.post("/bulk-delete", status_code=204)
async def bulk_delete_events(session: SessionDep, slug: str, data: EventBulkDelete):
    await event_service.bulk_delete_events(session, slug, data)


@router.get("/{event_id}", response_model=EventResponse)
async def get_event(session: SessionDep, slug: str, event_id: uuid.UUID):
    return await event_service.get_event(session, slug, event_id)


@router.patch("/{event_id}", response_model=EventResponse)
async def update_event(session: SessionDep, slug: str, event_id: uuid.UUID, data: EventUpdate):
    return await event_service.update_event(session, slug, event_id, data)


@router.patch("/{event_id}/move", response_model=EventResponse)
async def move_event(session: SessionDep, slug: str, event_id: uuid.UUID, data: EventMove):
    return await event_service.move_event(session, slug, event_id, data)


@router.delete("/{event_id}", status_code=204)
async def delete_event(session: SessionDep, slug: str, event_id: uuid.UUID):
    await event_service.delete_event(session, slug, event_id)
